mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp

```diff
@@ -306,6 +307,73 @@ emulatedVectorLoad(OpBuilder &rewriter, Location loc, Value base,
       newLoad);
 }
 
+/// Atomically store a subbyte-sized value to memory, with a mask.
+static void atomicStore(OpBuilder &rewriter, Location loc,
+                        TypedValue<MemRefType> emulatedMemref,
+                        Value emulatedIndex, TypedValue<VectorType> value,
+                        Value mask, int64_t scale) {
+  auto atomicOp = rewriter.create<memref::GenericAtomicRMWOp>(
+      loc, emulatedMemref, ValueRange{emulatedIndex});
+  OpBuilder builder =
+      OpBuilder::atBlockEnd(atomicOp.getBody(), rewriter.getListener());
+  Value origValue = atomicOp.getCurrentValue();
+
+  // i8 -> vector type <1xi8> then <1xi8> -> <scale x i.>
+  auto oneVectorType = VectorType::get({1}, origValue.getType());
+  auto fromElem = builder.create<vector::FromElementsOp>(loc, oneVectorType,
+                                                         ValueRange{origValue});
+  auto vectorBitCast =
+      builder.create<vector::BitCastOp>(loc, value.getType(), fromElem);
+
+  auto select =
+      builder.create<arith::SelectOp>(loc, mask, value, vectorBitCast);
+  auto bitcast2 = builder.create<vector::BitCastOp>(loc, oneVectorType, select);
+  auto extract = builder.create<vector::ExtractOp>(loc, bitcast2, 0);
+  builder.create<memref::AtomicYieldOp>(loc, extract.getResult());
```

Comment:
IIUC, this path is for the non data-contention cases. In this case, can we use few `memref::AtomicRMWOp` ops to get rid of `GenericAtomicRMWOp`? (I assumed that the `GenericAtomicRMWOp` is more expensive.)

Here is an example from `memref.store`. It uses `arith::AtomicRMWKind::andi` to clear the bits, and `arith::AtomicRMWKind::ori` to write the bits.

https://github.com/llvm/llvm-project/blob/d5032b9f4b6aa415e7fd39701f29edb93028d8b3/mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp#L435-L450

Reply:
It is a little bit difficult to change as it is tightly connected with `rmwStore`. Also, I do think that such optimization can be relatively easily handled on LLVM level, so I would say let's delegate to LLVM to do the hardwork. And if it does not optimize it then we can add a new pattern in LLVM to handle that.

My idea is that on MLIR level we can keep the semantics of the transformation readable and do not worry about detailed instruction level optimizations. Let me know what you think!